File: js/src/hash.test.ts
import { Hash, hash, hashCombine, HASH_SIZE_BYTES } from './hash';
import { describe, test, expect } from '@jest/globals';

describe('Hash', () => {
  test('hash size should be correct', () => {
    const data = 'TOPAY-Z512 test data';
    const hashValue = Hash.new(data);
    expect(hashValue.getBytes().length).toBe(HASH_SIZE_BYTES);
  });

  test('hash should be deterministic', () => {
    const data = 'TOPAY-Z512 test data';
    const hash1 = Hash.new(data);
    const hash2 = Hash.new(data);
    expect(hash1.equals(hash2)).toBe(true);
    expect(hash1.toHex()).toBe(hash2.toHex());
  });

  test('different inputs should produce different hashes', () => {
    const data1 = 'TOPAY-Z512 test data 1';
    const data2 = 'TOPAY-Z512 test data 2';
    const hash1 = Hash.new(data1);
    const hash2 = Hash.new(data2);
    expect(hash1.equals(hash2)).toBe(false);
    expect(hash1.toHex()).not.toBe(hash2.toHex());
  });

  test('combine should be different from concatenation', () => {
    const data1 = 'TOPAY-Z512';
    const data2 = 'test data';
    
    // Combined hash
    const combined = Hash.combine(data1, data2);
    
    // Concatenated hash
    const concatenated = Hash.new(data1 + data2);
    
    expect(combined.equals(concatenated)).toBe(false);
    expect(combined.toHex()).not.toBe(concatenated.toHex());
  });

  test('hex conversion should be reversible', () => {
    const data = 'TOPAY-Z512 hex conversion test';
    const hashValue = Hash.new(data);
    const hex = hashValue.toHex();
    const hashFromHex = Hash.fromHex(hex);
    
    expect(hashValue.equals(hashFromHex)).toBe(true);
    expect(hashValue.toHex()).toBe(hashFromHex.toHex());
  });

  test('should reject invalid hex strings', () => {
    // Invalid length
    expect(() => {
      Hash.fromHex('invalid');
    }).toThrow(/Invalid hex length/);
    
    // Invalid characters
    expect(() => {
      Hash.fromHex('zz' + '0'.repeat(HASH_SIZE_BYTES * 2 - 2));
    }).toThrow(/Invalid hex string/);
  });

  test('convenience functions should work correctly', () => {
    const data = 'TOPAY-Z512 test data';
    const data1 = 'TOPAY-Z512';
    const data2 = 'test data';
    
    // Test hash function
    const hashBytes = hash(data);
    expect(hashBytes.length).toBe(HASH_SIZE_BYTES);
    expect(hashBytes).toEqual(Hash.new(data).getBytes());
    
    // Test hashCombine function
    const combineBytes = hashCombine(data1, data2);
    expect(combineBytes.length).toBe(HASH_SIZE_BYTES);
    expect(combineBytes).toEqual(Hash.combine(data1, data2).getBytes());
  });

  test('should work with Uint8Array input', () => {
    const textData = 'TOPAY-Z512 test data';
    const binaryData = new TextEncoder().encode(textData);
    
    const hashFromText = Hash.new(textData);
    const hashFromBinary = Hash.new(binaryData);
    
    expect(hashFromText.equals(hashFromBinary)).toBe(true);
  });
});

File: go/pkg/hash/hash_test.go
package hash

import (
	"bytes"
	"testing"
)

func TestHashSize(t *testing.T) {
	data := []byte("TOPAY-Z512 test data")
	// Create hash and directly check its length without storing the variable
	if size := len(New(data)); size != HashSizeBytes {
		t.Errorf("Expected hash size %d, got %d", HashSizeBytes, size)
	}
}

func TestHashDeterministic(t *testing.T) {
	data := []byte("TOPAY-Z512 test data")
	hash1 := New(data)
	hash2 := New(data)

	if hash1 != hash2 {
		t.Errorf("Expected identical hashes for identical inputs")
	}
}

func TestHashDifferentInputs(t *testing.T) {
	data1 := []byte("TOPAY-Z512 test data 1")
	data2 := []byte("TOPAY-Z512 test data 2")
	hash1 := New(data1)
	hash2 := New(data2)

	if hash1 == hash2 {
		t.Errorf("Expected different hashes for different inputs")
	}
}

func TestHashCombine(t *testing.T) {
	data1 := []byte("TOPAY-Z512")
	data2 := []byte("test data")

	// Combined hash
	combined := Combine(data1, data2)

	// Concatenated hash
	concatenated := append([]byte{}, data1...)
	concatenated = append(concatenated, data2...)
	concatHash := New(concatenated)

	// These should be different
	if combined == concatHash {
		t.Errorf("Expected combined hash to be different from concatenated hash")
	}
}

func TestHexConversion(t *testing.T) {
	data := []byte("TOPAY-Z512 hex conversion test")
	hash := New(data)
	hexStr := hash.String()

	hash2, err := FromHex(hexStr)
	if err != nil {
		t.Errorf("Error converting from hex: %v", err)
	}

	if hash != hash2 {
		t.Errorf("Expected identical hashes after hex conversion")
	}
}

func TestInvalidHex(t *testing.T) {
	// Test invalid length
	_, err := FromHex("invalid")
	if err == nil {
		t.Errorf("Expected error for invalid hex length")
	}

	// Test invalid characters
	_, err = FromHex("zz" + repeat("00", HashSizeBytes-1))
	if err == nil {
		t.Errorf("Expected error for invalid hex characters")
	}
}

// Helper function to repeat a string
func repeat(s string, count int) string {
	var result string
	for i := 0; i < count; i++ {
		result += s
	}
	return result
}

func TestBytes(t *testing.T) {
	data := []byte("TOPAY-Z512 bytes test")
	hash := New(data)
	hashBytes := hash.Bytes()

	if !bytes.Equal(hashBytes, hash[:]) {
		t.Errorf("Expected Bytes() to return the hash bytes")
	}
}
